### PR TITLE
add power devices menu

### PR DESCRIPTION
**Pull request by kevdliu**
This PR adds a power devices sub-menu to the printer tuning menu that allows users to turn on/off power devices exposed via moonraker. Examples of power devices include smart plugs that the printer is connected to or external fans etc. 

I'm pretty new to LGVL so appreciate any feedback you have. The only thing missing is the icon for the menu button as I'm unsure how to generate an icon for the button. I'm using the graph icon as a placeholder for now. 

Capture
2Capture

### CODE REVIEW

**Comment by kevdliu** [comment]
I'm trying to cross compile the binary for my Creality Ender V3 KE (which has the same chipset as K1) but keep running into this error:
```
/home/kevdliu/mips-gcc720-glibc229/bin/../lib/gcc/mips-linux-gnu/7.2.0/../../../../mips-linux-gnu/bin/ld: skipping incompatible spdlog/build/libspdlog.a when searching for -l:libspdlog.a
/home/kevdliu/mips-gcc720-glibc229/bin/../lib/gcc/mips-linux-gnu/7.2.0/../../../../mips-linux-gnu/bin/ld: skipping incompatible spdlog/build/libspdlog.a when searching for -l:libspdlog.a
/home/kevdliu/mips-gcc720-glibc229/bin/../lib/gcc/mips-linux-gnu/7.2.0/../../../../mips-linux-gnu/bin/ld: cannot find -l:libspdlog.a
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:132: default] Error 1
make[1]: Leaving directory '/home/kevdliu/Downloads/guppyscreen'
make: *** [Makefile:162: build] Error 2
```

I've downloaded the mips sdk and included it in my path. Building natively for x86 works fine but when I set `export CROSS_COMPILE=mips-linux-gnu-` it throws the above error

**Comment by ballaswag** [comment]
Thanks for the PR! See https://github.com/ballaswag/guppyscreen/pull/62/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R105 to get around that spdlog issue.

For icons, I use the lvgl image convert tool to convert svg into c arrays (https://github.com/lvgl/lv_img_conv):
./lv_img_conv.js logo_lvgl.png -f -c CF_TRUE_COLOR_ALPHA

The icon set I'm using are from https://pictogrammers.com/library/mdi/ (I'm out of the zbolt icons since that set is custom). There are also 2 sizes because I didn't bother trying to figure out how to dynamically scale them in lvgl.

The default set is 64x64, and the small screen ones are 46x46.

**Comment by kevdliu** [comment]
> Thanks for the PR! See https://github.com/ballaswag/guppyscreen/pull/62/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R105 to get around that spdlog issue.
> 
> For icons, I use the lvgl image convert tool to convert svg into c arrays (https://github.com/lvgl/lv_img_conv): ./lv_img_conv.js logo_lvgl.png -f -c CF_TRUE_COLOR_ALPHA
> 
> The icon set I'm using are from https://pictogrammers.com/library/mdi/ (I'm out of the zbolt icons since that set is custom). There are also 2 sizes because I didn't bother trying to figure out how to dynamically scale them in lvgl.
> 
> The default set is 64x64, and the small screen ones are 46x46.

Thanks for the spdlog suggestion! It's working now. I've updated the PR with a new icon for the power devices menu. Unsure what to do about ZBolt though.

**Comment by kevdliu** [comment]
New icon:
Screenshot from 2024-04-15 21-57-22

File: src/power_panel.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    lv_obj_add_state(power_device_toggle, LV_STATE_CHECKED);
+  } else {
+    lv_obj_clear_state(power_device_toggle, LV_STATE_CHECKED);
+  }
+}
+
+void PowerPanel::create_devices(json &j) {
+  std::lock_guard<std::mutex> lock(lv_lock);
+
+  if (j.contains("result")) {
```
Comment (ballaswag) [inline src/power_panel.cpp:71]: probably need to clear the devices map here for when the printer re-initializes, so if any devices removed physically will also get removed from in the UI.

File: src/printertune_panel.cpp [hunk 2]
```diff
@@ -37,6 +41,11 @@ PrinterTunePanel::PrinterTunePanel(KWebSocketClient &c, std::mutex &l, lv_obj_t
   , limits_btn(cont, &limit_img, "Limits", &PrinterTunePanel::_handle_callback, this)
   , tmc_tune_btn(cont, &motor_img, "TMC Autotune", &PrinterTunePanel::_handle_callback, this)
   , tmc_status_btn(cont, &chart_img, "TMC Metrics", &PrinterTunePanel::_handle_callback, this)
+#ifndef ZBOLT
+  , power_devices_btn(cont, &power_devices_img, "Power Devices", &PrinterTunePanel::_handle_callback, this)
+#else
+  , power_devices_btn(cont, &print, "Power Devices", &PrinterTunePanel::_handle_callback, this)
```
Comment (ballaswag) [inline src/printertune_panel.cpp:47]: I just start using material icons for zbolt too when there's no zbolt equivalent.

**Comment by ballaswag** [comment]
Thanks for the feature! Will merge and let it bake in nightly. Some minor comments, if you don't a chance to fix, I'll see to them when I get some time.

Another thing is I'll probably move this icon into a different section and leave the tuning panel for tuning stuff. For now it's good enough.

File: src/power_panel.cpp (hunk 1 above)
Comment (kevdliu) [reply src/power_panel.cpp:71]: Thanks for the suggestion. Do you know how I can remove existing UI widgets in LVGL? Simply clearing the map still leaves the UI widget on screen and causes duplicate entries.

**Pull request merged**